include/xtensor/xset_operation.hpp

```diff
@@ -24,6 +24,58 @@
 
 namespace xt
 {
+    namespace detail
+    {
+
+        template<typename T>
+        struct has_begin_end
+        {
+            private:
+
+                template<typename C> static constexpr auto test(void*)
+                    -> decltype(std::declval<C>().begin(), std::declval<C>().end(), std::true_type())
+                {
+                    return std::true_type();
+                }
+
+                template<typename C> static constexpr std::false_type test(...)
+                {
+                    return std::false_type();
+                }
+
+            public:
+
+                static constexpr const bool value = std::is_same<decltype(test<T>(0)), std::true_type>::value;
+        };
```

Comment:
A more concise way for checking the existence of methods is to rely on template default parameters and template specialization:

```cpp
template <class T, class = void>
struct has_begin_end : std::false_type
{
};

template <class T>
struct has_begin_end<T, void_t<decltype(std::declval<T>().begin(), std::declval<T>().end())>>
    : std::true_type
{
};
```

Reply:
Actually, `xutils.hpp` already provides the `has_iterator_interface` metafunction. It only checks for the `begin` method, so maybe it would be worth checking for the end method too.

Reply:
Nice! I added `is_iterator` to it too